Content.Server/TapeRecorder/TapeRecorderSystem.cs

```diff
@@ -0,0 +1,162 @@
+        }
+
+        return true;
+    }
+
+    protected override bool ProcessRewindingTapeRecorder(Entity<TapeRecorderComponent> tapeRecorder, float frameTime)
+    {
+        if (!base.ProcessRewindingTapeRecorder(tapeRecorder, frameTime))
+        {
+            Stop(tapeRecorder, changeMode: true);
+            return false;
+        }
+
+        return true;
+    }
+
+    /// <summary>
+    /// Given a time range, play all messages on a tape within said range
+    /// Split into this system as shared does not have _chatSystem access
+    /// </summary>
+    protected override void ReplayMessagesInSegment(Entity<TapeRecorderComponent> tapeRecorder, TapeCassetteComponent tapeCassetteComponent, float segmentStart, float segmentEnd)
+    {
+        EnsureComp<VoiceMaskComponent>(tapeRecorder, out var voiceMaskComponent);
+
+        foreach (var messageToBeReplayed in tapeCassetteComponent.RecordedData.Where(x => x.Timestamp > tapeCassetteComponent.CurrentPosition && x.Timestamp <= segmentEnd))
+        {
+            //Change the voice to match the speaker
+            if (voiceMaskComponent != null)
+                voiceMaskComponent.VoiceName = messageToBeReplayed.Name ?? messageToBeReplayed.DefaultName;
+            //Play the message
+            _chatSystem.TrySendInGameICMessage(tapeRecorder, messageToBeReplayed.Message, InGameICChatType.Speak, false);
+        }
+    }
+
+    /// <summary>
+    /// Whenever someone speaks within listening range, record it to tape
+    /// </summary>
+    private void OnListen(Entity<TapeRecorderComponent> tapeRecorder, ref ListenEvent args)
+    {
+        if (tapeRecorder.Comp.Mode != TapeRecorderMode.Recording || !tapeRecorder.Comp.Active)
+            return;
+
+        if (!TryGetTapeCassette(tapeRecorder, out var cassette))
+            return;
+
+        //Handle someone using a voice changer
+        var nameEv = new TransformSpeakerNameEvent(args.Source, Name(args.Source));
+        RaiseLocalEvent(args.Source, nameEv);
+
+        //Add a new entry to the tape
+        cassette.Comp.Buffer.Add(new TapeCassetteRecordedMessage(cassette.Comp.CurrentPosition, nameEv.Name, args.Message));
+    }
+
+
+    /// <summary>
+    /// Start playback if we are not already playing, ensure we have a voice mask component so the name is correctly shown through radios
+    /// </summary>
+    protected override bool StartPlayback(Entity<TapeRecorderComponent> tapeRecorder, EntityUid? user = null)
+    {
+        if (base.StartPlayback(tapeRecorder, user))
+        {
+            //Server only component
+            EnsureComp<VoiceMaskComponent>(tapeRecorder);
+            return true;
+        }
+
+        return false;
+    }
+
+    private void OnPrintMessage(EntityUid uid, TapeRecorderComponent tapeRecorder, PrintTapeRecorderMessage args)
+    {
+        var text = new StringBuilder();
+        var paper = EntityManager.SpawnEntity(tapeRecorder.PaperPrototype, Transform(uid).Coordinates);
```

Comment:
This should use regular `Spawn`. Also it needs to not be spamable for infinite paper somehow.

Reply:
> This should use regular `Spawn`. Also it needs to not be spamable for infinite paper somehow.

UI print button have 4sec CD for spam preventing

Reply:
You need a server-side check for the cooldown. All forms of validation like length checks, rate limiting, needs to be done on the server.

Reply:
> You need a server-side check for the cooldown. All forms of validation like length checks, rate limiting, needs to be done on the server.

Added server side cooldown check